synapse/appservice/__init__.py

```diff
@@ -172,12 +172,24 @@ async def _matches_user_in_member_list(
         Returns:
             True if this service would like to know about this room.
         """
-        member_list = await store.get_users_in_room(
+        # We can use `get_local_users_in_room(...)` here because an application service
+        # can only be interested in local users of the server it's on (ignore any remote
+        # users that might match the user namespace regex).
```

Comment:
We have to rip the band-aid off at some point whether it before or after [MSC3905](https://github.com/matrix-org/matrix-spec-proposals/pull/3905) merges. People relying on the old behavior will probably have to find out the hard way that it doesn't work anymore. I can't think of a way to add a warning for this scenario that wouldn't generate a ton of noise (false-positives) 🤔

We could detect a very specific case where someone explicitly defines the `@.*:hostname` part of the regex that doesn't match the server hostname.

Reply:
I wonder if it's worth adding an entry to the [upgrade notes](https://matrix-org.github.io/synapse/latest/upgrade)?

Reply:
Added something to the upgrade notes 👍